### ISSUE TITLE
Using with Multiple GPUs

### ISSUE THREAD

**Issue by ShivaKothuru** [nianticlabs/monodepth2#82]
Hello,

I would appreciate if you help me with following issue.
Is this training can be done using multiple GPUs? If yes, refering to  [(https://github.com/nianticlabs/monodepth2/issues/5)](url) changed both files "depth and pose decoder". However, it still continue to eat the resources from single GPU.

Please suggest any changes if required.

Pose_Decoder:

        setattr(self, "squeeze", nn.Conv2d(self.num_ch_enc[-1], 256, 1))
        
        setattr(self, "pose_{}_".format(0), nn.Conv2d(num_input_features * 256, 256, 3, stride, 1))
        
        setattr(self, "pose_{}_".format(1), nn.Conv2d(256, 256, 3, stride, 1))
       
        setattr(self, "pose_{}_".format(2), nn.Conv2d(256, 6 * num_frames_to_predict_for, 1))

        self.relu = nn.ReLU()

        #self.net = nn.ModuleList(list(self.convs.values()))

    def forward(self, input_features):
        last_features = [f[-1] for f in input_features]

        #cat_features = [self.relu(self.convs["squeeze"](f)) for f in last_features]
        cat_features = [self.relu(getattr(self, "squeeze")(f)) for f in last_features]

Depth_decoder:

        # decoder
        #self.convs = OrderedDict()
        for i in range(4, -1, -1):
            # upconv_0
            num_ch_in = self.num_ch_enc[-1] if i == 4 else self.num_ch_dec[i + 1]
            num_ch_out = self.num_ch_dec[i]
            #self.convs[("upconv", i, 0)] = ConvBlock(num_ch_in, num_ch_out)
            setattr(self, "upconv_{}_0".format(i), ConvBlock(num_ch_in, num_ch_out))

            # upconv_1
            num_ch_in = self.num_ch_dec[i]
            if self.use_skips and i > 0:
                num_ch_in += self.num_ch_enc[i - 1]
            num_ch_out = self.num_ch_dec[i]
            #self.convs[("upconv", i, 1)] = ConvBlock(num_ch_in, num_ch_out)
            setattr(self, "upconv_{}_1".format(i), ConvBlock(num_ch_in, num_ch_out))

        for s in self.scales:
            #self.convs[("dispconv", s)] = Conv3x3(self.num_ch_dec[s], self.num_output_channels)
            setattr(self, "dispconv_{}_".format(s), Conv3x3(self.num_ch_dec[s], self.num_output_channels)) 

        #self.decoder = nn.ModuleList(list(self.convs.values()))
        self.sigmoid = nn.Sigmoid()

    def forward(self, input_features):
        self.outputs = {}

        # decoder
        x = input_features[-1]
        for i in range(4, -1, -1):
            #x = self.convs[("upconv", i, 0)](x)
            x = getattr(self, "upconv_{}_0".format(i))(x)
            x = [upsample(x)]
            if self.use_skips and i > 0:
                x += [input_features[i - 1]]
            x = torch.cat(x, 1)
            #x = self.convs[("upconv", i, 1)](x)
            x = getattr(self, "upconv_{}_1".format(i))(x)
            
            if i in self.scales:
                #self.outputs[("disp", i)] = self.sigmoid(self.convs[("dispconv", i)](x))
                self.outputs[("disp", i)] = self.sigmoid(getattr(self, "dispconv_{}_".format(i))(x))
                #setattr(self.outputs, "disp_{}_".format(i), self.sigmoid(getattr(self, "dispconv_{}_".format(i))(x)))

**Comment by daniyar-niantic** [comment]
Hi @ShivaKothuru.

The changes that you have written allow pytorch to know how to place the model onto multiple GPUs. You still have to tell pytorch to use multiple gpus. One way is to parallelise the model with [nn.DataParallel](https://pytorch.org/docs/0.4.1/nn.html#torch.nn.DataParallel).
